### DISCUSSION TITLE
Question: mac layout change watcher

### DISCUSSION THREAD

**Discussion by alex-enchi** [zsa/kontroll#12] [Q&A]
Hi all

Do you know is there a good method to be notified on layout change on macos, aside of checking with interval for
`defaults read ~/Library/Preferences/com.apple.HIToolbox.plist AppleSelectedInputSources`

Ideally I want to set specific layer on when changing layout and interval does not sound like a best option

Best

**Comment by fdidron** [comment]
I'm not super familiar with macOS, but I think this will require some Swift / ObjectiveC code to subscribe to system's notifications.

Wouldn't it be easier to have a script (Apple script or any automation tool) that you invoke with a key shortcut that:
1. Sets or cycles the input method
2. Read the value of AppleSelectedInputSources
3. Sets the layout accordingly to the value with kontroll

**Comment by alex-enchi** [comment]
I thought about it as well, but it means that layout must be changed only through shortcut.
Recently I tried option with listening for file changes of that plist file, but apparently file is not updated right away.

**Reply by jcrawford** [reply]
@alex-enchi You can use AppleScript but you would have to use polling to determine if the value has changed.  You could use the on change handler but this would only be fired when the value is changed by the user and not when the system changes the value.

A simple polling script would look something similar to this.

```AppleScript
on run
  tell application "System Events"
    set previousInputSource to (get AppleSelectedInputSources)
  end tell

  -- Loop to continuously check for changes
  repeat
    tell application "System Events"
      set currentInputSource to (get AppleSelectedInputSources)
    end tell

    if currentInputSource is not equal to previousInputSource then
      -- Execute the shell script to interact with the keyboard
      do shell script "/path/to/your/update-layout.sh"
      
      -- Save the new source for future use
      set previousInputSource to currentInputSource
    end if

    delay 1  -- Check every second (adjust as needed)
  end repeat
end run
```

If your bash script is simply executing a `kontroll` command you could do all of the work in the apple script itself and not have to call the bash script. `do shell script "kontroll set-layer --layer <INDEX>"`

When it comes to making your applescript run continuously in the background `launchd` is your friend, you can see how it is done here: https://apple.stackexchange.com/q/192600

**Comment by MaxFreedomPollard** [comment]
There is an event for this, so you don't have to poll the plist at all. The cleanest way to get it without writing your own Swift/ObjC is Hammerspoon: `hs.keycodes.inputSourceChanged(fn)` runs your callback whenever the input source or layout changes, and `hs.keycodes.currentSourceID()` tells you which one it changed to. Then you just shell out to `kontroll`.

```lua
-- ~/.hammerspoon/init.lua
local KONTROLL = "/opt/homebrew/bin/kontroll"  -- adjust to `which kontroll`

-- map macOS input-source ids to the kontroll layer you want on each
local layerForSource = {
  ["com.apple.keylayout.US"]     = 0,
  ["com.apple.keylayout.Dvorak"] = 1,
}

hs.keycodes.inputSourceChanged(function()
  local sid = hs.keycodes.currentSourceID()
  local layer = layerForSource[sid]
  if layer then
    -- replace <set-layer> with your actual kontroll command
    hs.execute(KONTROLL .. " <set-layer> " .. layer, true)  -- true = load your shell env for PATH
  end
end)
```

To find the exact source id for each of your layouts, switch to the layout and run `hs.keycodes.currentSourceID()` in the Hammerspoon console (there's also `hs.keycodes.currentLayout()` for the human-readable name).

Why this is better than the interval or the plist watch you tried: `inputSourceChanged` is driven by the system's Text Input Source change notification, so it fires immediately on the change. It doesn't depend on when `com.apple.HIToolbox.plist` gets flushed to disk, which is exactly the lag you ran into.

If you'd rather not depend on Hammerspoon, the underlying primitive is that same notification: a small Swift launch agent that adds a `DistributedNotificationCenter` observer for the selected-keyboard-input-source-changed notification (`kTISNotifySelectedKeyboardInputSourceChanged`) and shells out to kontroll does the identical job. Hammerspoon just hands you the event, the current-source lookup, and the shell-out in about ten lines.
